public/data.js

```diff
@@ -27,19 +27,36 @@ class DataController {
             // Update data every time it changes in firestore
             DataController.updateData(res, activeProjects);
             onChange(activeProjects);
+        }, error => {
+            if (error.code == "permission-denied") {
+                firebase.auth().onAuthStateChanged(user => { 
+                    if (user) {
+                        // Error message - Alert Missing or insufficient permissions.
+                        alert(`${error.message} 
+                        \r Please use AVF email or liaise with Africa's Voices Foundation (AVF) to give you AVF email`) 
+                        if (user.email.match(".*@africasvoices.org$")) {
+                            alert(`${error.message} 
+                            \r Please liaise with Africa's Voices Foundation (AVF) to activate your email`) 
+                        }
+                    }
+                })
+                window.location.replace("auth.html")
+            } else {
+                console.log(error);
+            }     
         });
     }
 
     static watchCodingProgress(onChange) {
         return mediadb.doc("metrics/coda").onSnapshot(res => {
             onChange(res.data());
-        });
+        }, error => console.log(error));
```

Comment:
do we still need this?

Reply:
Yes, after handling the error, the listener will not receive any more events, and there is no need to detach your listener